File: Backend/src/main/java/aor/paj/projecto4/service/ClientsService.java
package aor.paj.projecto4.service;

import java.util.List;

import aor.paj.projecto4.bean.ClientsBean;
import aor.paj.projecto4.dto.ClientsDTO;
import jakarta.inject.Inject;
import jakarta.validation.Valid;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.DELETE;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.HeaderParam;
import jakarta.ws.rs.PATCH;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.PUT;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.PathParam;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.QueryParam;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;

/**
 * SERVIÇO REST: ClientsService
 * ----------------------------
 * DESCRIÇÃO: Controlador JAX-RS responsável pela gestão de clientes.
 * ARQUITETURA: Atua como a camada de apresentação da API, recebendo os pedidos HTTP,
 * delegando a segurança ao 'verifier' e a lógica de negócio ao 'ClientsBean'.
 */
@Path("/clients")
public class ClientsService {

    @Inject
    ClientsBean clientsBean;

    @Inject
    UserVerificationBean verifier;

    // =========================================================================
    // --- 1. OPERAÇÕES DE UTILIZADOR E DONO ---
    // =========================================================================

    /**
     * CRIAÇÃO DE CLIENTE (RESTful POST)
     * O uso de @Valid garante que a validação do DTO ocorra antes de atingir o Bean.
     */
    @POST
    @Path("/")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public Response addClient(@HeaderParam("token") String token, @Valid ClientsDTO clientDto) {
        // SEGURANÇA: Garante que apenas utilizadores com sessão ativa podem criar.
        verifier.verifyUser(token);

        ClientsDTO createdClient = clientsBean.addClient(token, clientDto);

        // BOAS PRÁTICAS REST: Retorna 201 Created quando um novo recurso é gerado.
        return Response.status(201).entity(createdClient).build();
    }

    /**
     * LISTAGEM DINÂMICA (POLIMORFISMO DE DADOS)
     * O comentário indica que o Bean decide internamente o que devolver com base no token.
     */
    @GET
    @Path("/")
    @Produces(MediaType.APPLICATION_JSON)
    public Response listClients(
            @HeaderParam("token") String token,
            @QueryParam("userId") Long userId // Permite filtragem opcional via URL
    ) {
        verifier.verifyUser(token);
        // LÓGICA DE NEGÓCIO: O Bean trata a lógica Admin vs User internamente
        List<ClientsDTO> clients = clientsBean.listClients(token, userId);
        return Response.ok(clients).build();
    }

    /**
     * ATUALIZAÇÃO DE CLIENTE
     */
    @PUT
    @Path("/{id:[0-9]+}")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public Response updateClient(@PathParam("id") Long clientId,
                                 @HeaderParam("token") String token,
                                 @Valid ClientsDTO clientDto) {
        // SEGURANÇA (RBAC): Valida simultaneamente se o token é válido E se o utilizador
        // tem permissão para editar este recurso específico (é o Dono ou é Admin).
        verifier.verifyOwnershipOrAdmin(token, clientId);

        ClientsDTO updatedClient = clientsBean.editClient(clientId, clientDto);
        return Response.ok(updatedClient).build();
    }

    /**
     * SOFT DELETE DE CLIENTE (Lixeira)
     * -------------------------------------------------------------------------
     * CORREÇÃO DE ROTEAMENTO: A adição de `:[0-9]+` garante que o servidor JAX-RS
     * este endpoint só aceita números, evitando colisões com rotas
     * estáticas como "/me-trash" ou "/trash".
     */
    @DELETE
    @Path("/{id:[0-9]+}")
    public Response softDelete(@HeaderParam("token") String token, @PathParam("id") Long id) {
        // SEGURANÇA (RBAC): Valida se o utilizador é o dono do registo ou um Administrador
        verifier.verifyOwnershipOrAdmin(token, id);

        clientsBean.softDeleteClient(id);

        // BOAS PRÁTICAS REST: 204 No Content indica sucesso numa operação de eliminação
        return Response.noContent().build();
    }

    /**
     * RESTAURO DE CLIENTE
     * -------------------------------------------------------------------------
     * Utiliza PATCH porque estamos a realizar uma atualização parcial (apenas a flag softDelete).
     * Regex `:[0-9]+` aplicada ao PathParam.
     */
    @PATCH
    @Path("/{id:[0-9]+}/restore")
    @Produces(MediaType.APPLICATION_JSON) // Garante que o cabeçalho HTTP devolve o formato correto
    public Response restoreClient(@HeaderParam("token") String token, @PathParam("id") Long id) {
        // SEGURANÇA (RBAC): Verificação de posse ou privilégios globais
        verifier.verifyOwnershipOrAdmin(token, id);

        // Guarda o DTO retornado pelo Bean após a alteração de estado
        ClientsDTO restoredClient = clientsBean.restoreClient(id);

        // Retorna o objeto completo para que o Frontend possa atualizar a UI sem fazer novos pedidos
        return Response.ok(restoredClient).build();
    }

    /**
     * Endpoint: Lixeira exclusiva do próprio utilizador (User Normal)
     * Rota React: api("/clients/me-trash", "GET")
     */
    @GET
    @Path("/me-trash")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getMyTrash(@HeaderParam("token") String token) {
        // 1. Apenas exige que seja um utilizador válido
        verifier.verifyUser(token);

        // 2. LÓGICA INTELIGENTE: Como ele não é Admin,
        // o teu Bean vai ignorar o 'null' e usar o ID do próprio utilizador.
        List<ClientsDTO> trash = clientsBean.listDeletedClientsDTO(token, null);

        return Response.ok(trash).build();
    }


    // =========================================================================
    // --- 2. OPERAÇÕES EXCLUSIVAS DE ADMINISTRADOR ---
    // =========================================================================

    /**
     * HARD DELETE (Remoção Física)
     * -------------------------------------------------------------------------
     * Exclusivo para operações de limpeza permanente da base de dados.
     * Regex `:[0-9]+` isola o parâmetro dinâmico das rotas de sistema.
     */
    @DELETE
    @Path("/{id:[0-9]+}/permanent")
    public Response permanentDelete(@HeaderParam("token") String token, @PathParam("id") Long id) {
        // 1. SEGURANÇA ESTRITA: Apenas o perfil Administrador tem acesso a eliminações físicas
        verifier.verifyAdmin(token);

        // 2. Chama o método de execução destrutiva no Bean
        clientsBean.permanentDeleteClient(id);

        // 3. Retorna 204 No Content
        return Response.noContent().build();
    }

    /**
     * AÇÃO EM LOTE: Desativar todos os clientes de um utilizador.
     */
    @PATCH
    @Path("/user/{userId}/status/deactivate-all")
    public Response softDeleteAllUserClients(
            @HeaderParam("token") String token,
            @PathParam("userId") Long userId) {
        verifier.verifyAdmin(token);
        // LÓGICA DE PERFORMANCE: Bulk Update em vez de iterar cliente a cliente.
        clientsBean.softDeleteAllClientsByUser(userId);
        return Response.ok().build();
    }

    /**
     * LIXEIRA POR UTILIZADOR (Visão de Admin)
     */
    @GET
    @Path("/user/{userId}/trash")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getDeletedClientsByUserId(
            @HeaderParam("token") String token,
            @PathParam("userId") Long userId) {
        verifier.verifyAdmin(token);
        // Atualizado para usar o novo método dinâmico com filtro
        List<ClientsDTO> clients = clientsBean.listDeletedClientsDTO(token, userId);
        return Response.ok(clients).build();
    }

    /**
     * AÇÃO EM LOTE: Restaurar todos os clientes de um utilizador.
     */
    @PATCH
    @Path("/user/{userId}/status/activate-all")
    public Response restoreAllUserClients(
            @HeaderParam("token") String token,
            @PathParam("userId") Long userId) {
        verifier.verifyAdmin(token);
        clientsBean.unSoftDeleteAllClientsByUser(userId); // Agora usa Bulk Update
        return Response.ok().build();
    }

    /**
     * AÇÃO DE LIMPEZA GERAL (Hard Delete em Lote)
     */
    @DELETE
    @Path("/user/{userId}/trash")
    public Response emptyTrash(
            @PathParam("userId") Long userId,
            @HeaderParam("token") String token) {
        verifier.verifyAdmin(token);
        clientsBean.emptyTrash(userId);
        return Response.noContent().build();
    }

    /**
     * CRIAÇÃO DE CLIENTE POR PROCURAÇÃO
     * O Admin cria o cliente, mas atribui-o a outro utilizador.
     */
    @POST
    @Path("/user/{userId}")
    public Response createClientForUser(
            @PathParam("userId") Long userId,
            @Valid ClientsDTO dto,
            @HeaderParam("token") String token) {
        verifier.verifyAdmin(token);
        ClientsDTO created = clientsBean.createClientForUser(userId, dto);
        return Response.status(Response.Status.CREATED).entity(created).build();
    }

    /**
     * LIXEIRA GLOBAL (Visão de Admin de todos os apagados)
     */
    @GET
    @Path("/trash")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getAllDeletedClients(@HeaderParam("token") String token) {
        verifier.verifyAdmin(token);
        // Chama a lixeira global sem filtro de utilizador
        List<ClientsDTO> clients = clientsBean.listAllDeletedClients();
        return Response.ok(clients).build();
    }
}

File: Frontend/src/services/clientsService.js
import api from "./api";

/**
 * SERVIÇO: clientsService
 * ----------------------
 * DESCRIÇÃO: Encapsula todas as chamadas à API JAX-RS para a entidade Client.
 * FUNCIONALIDADE: Implementa a lógica de negócio de clientes, incluindo
 * filtragem por utilizador, gestão de lixeira (Soft Delete) e remoção permanente.
 */
export const clientsService = {

  /**
   * ROTEADOR CENTRAL DE LISTAGEM (PADRÃO DE ESTRATÉGIA):
   * Este método decide qual o endpoint chamar baseando-se no Role e no estado (Lixeira ou Ativo).
   * Resolve a complexidade de rotas múltiplas num único ponto de entrada.
   */
  getClients: async (userRole, filters = {}) => {
    const { userId, showTrash } = filters;

    // --- 1. GESTÃO DE LIXEIRA (REGRA A9): ---
    if (showTrash) {
      if (userRole === "ADMIN") {
        if (userId) {
          // Admin consulta a lixeira de um colaborador específico (@GET /clients/user/{userId}/trash)
          return await api(`/clients/user/${userId}/trash`, "GET");
        } else {
          // Admin consulta a lixeira global do sistema (@GET /clients/trash)
          return await api("/clients/trash", "GET");
        }
      } else {
        // Utilizador Normal consulta apenas os seus próprios dados eliminados (@GET /clients/me-trash)
        return await api("/clients/me-trash", "GET");
      }
    }

    // --- 2. GESTÃO DE CLIENTES ATIVOS: ---
    const params = new URLSearchParams();
    if (userRole === "ADMIN" && userId) {
      // Injeta o ID do utilizador alvo como Query Parameter para filtragem no SQL.
      params.append("userId", userId);
    }

    const queryString = params.toString();
    const url = queryString ? `/clients?${queryString}` : "/clients";

    return await api(url, "GET");
  },

  /** * OPERAÇÕES CRUD (CREATE, READ, UPDATE, DELETE):
   * Mapeamento direto para os métodos do ClientResource no Java.
   */

  // Registo de novo cliente pelo próprio utilizador
  createClient: async (clientDto) => {
    return await api("/clients", "POST", clientDto);
  },

  // Criação de cliente atribuído a outrem (Exclusivo Admin)
  createClientForUser: async (userId, clientDto) => {
    return await api(`/clients/user/${userId}`, "POST", clientDto);
  },

  // Atualização de dados (Mapeia para @PUT /clients/{id})
  updateClient: async (id, clientDto) => {
    return await api(`/clients/${id}`, "PUT", clientDto);
  },

  /** * CICLO DE VIDA E ELIMINAÇÃO (REGRAS A9 e A14): */

  // Soft Delete: Marca o cliente como 'deleted' sem o remover da base de dados.
  softDeleteClient: async (id) => {
    return await api(`/clients/${id}`, "DELETE");
  },

  // Restaurar: Reverte o Soft Delete (Mapeia para @PATCH no Java conforme as boas práticas REST).
  restoreClient: async (id) => {
    return await api(`/clients/${id}/restore`, "PATCH");
  },

  // Hard Delete: Remoção definitiva do registo no PostgreSQL (Restrito a Admin).
  permanentDeleteClient: async (id) => {
    return await api(`/clients/${id}/permanent`, "DELETE");
  },

  /** * OPERAÇÕES EM MASSA (BULK ACTIONS):
   * Implementam a gestão rápida de grandes volumes de dados para Administradores.
   */

  // Desativa todos os clientes de um utilizador específico de uma só vez.
  softDeleteAllFromUser: async (userId) => {
    return await api(`/clients/user/${userId}/status/deactivate-all`, "PATCH");
  },

  // Restaura todos os clientes eliminados de um utilizador.
  restoreAllFromUser: async (userId) => {
    return await api(`/clients/user/${userId}/status/activate-all`, "PATCH");
  },

  // Limpa definitivamente a lixeira de um colaborador.
  emptyTrashByUserId: async (userId) => {
    return await api(`/clients/user/${userId}/trash`, "DELETE");
  },

  // Consulta administrativa da lixeira de todo o CRM.
  getAllTrash: async () => {
    return await api("/clients/trash", "GET");
  }
};